src/common/snippets/src/lowered/pass/extract_loop_invariants.cpp

```diff
@@ -0,0 +1,199 @@
+        }
+    }
+    if (!exp_out_ports.empty()) {
+        std::vector<ExpressionPort> new_ports;
+        inner_loop_info->update_loop_ports(exp_out_ports, new_ports);
+    }
+    // TODO: 142990.
+    // Need sort after update loop ports. There are possibility that all exprs are moved to outer loop.
+    if (!inner_loop_info->get_input_ports().empty() && !inner_loop_info->get_output_ports().empty()) {
+        loop_manager->sort_loop_ports(inner_loop_begin_pos, inner_loop_end_pos, inner_loop_id);
+    }
+}
+
+std::set<ExpressionPtr> get_loop_input_exprs(const std::vector<LoopPort>& loop_in_ports) {
+    std::set<ExpressionPtr> expr_set;
+    for (size_t i = 0; i < loop_in_ports.size(); ++i) {
+        expr_set.insert(loop_in_ports[i].expr_port->get_expr());
+    }
+    return expr_set;
+}
+
+bool extract_from_loop(const size_t& inner_loop_id, LinearIR& linear_ir) {
+    const auto& loop_manager = linear_ir.get_loop_manager();
+    bool status = false;
+    bool continue_to_extract = true;
+    const auto& inner_loop_info = loop_manager->get_loop_info<UnifiedLoopInfo>(inner_loop_id);
+    while (continue_to_extract) {
+        const auto& inner_loop_input_ports = inner_loop_info->get_input_ports();
+        const auto& potential_extractable_exprs = get_loop_input_exprs(inner_loop_input_ports);
+        bool expr_extracted = false;
+        for (const auto& port_expr : potential_extractable_exprs) {
+            if (is_extraction_applicable(port_expr, inner_loop_info)) {
+                status = true;
+                LinearIR::constExprIt inner_loop_begin_pos, inner_loop_end_pos;
+                std::tie(inner_loop_begin_pos, inner_loop_end_pos) =
+                    loop_manager->get_loop_bounds(linear_ir, inner_loop_id);
+                // extract scalar on inputs if there are
+                for (size_t i = 0; i < port_expr->get_input_count(); ++i) {
+                    auto parent = port_expr->get_input_port_connector(i)->get_source().get_expr();
+                    if (ov::is_type<snippets::op::Scalar>(parent->get_node())) {
+                        extract_expr(parent, linear_ir, inner_loop_begin_pos, inner_loop_end_pos);
+                    }
+                }
+                extract_expr(port_expr, linear_ir, inner_loop_begin_pos, inner_loop_end_pos);
+                update_loop_ports(port_expr, loop_manager, inner_loop_id, inner_loop_begin_pos, inner_loop_end_pos);
+                expr_extracted = true;
+                break;  // extracted and refreshed loop_input_ports. break potential_extractable_exprs loop, and go while() to start again.
+            }
+        }
+        if (inner_loop_input_ports.size() == 0 && inner_loop_info->get_output_ports().size() == 0) {
+            // become a empty(inner_loop_input_ports is ref) loop after extraction, let remove it from loop_manager
+            loop_manager->remove_loop_info(inner_loop_id);
+            continue_to_extract = false;
+            break;
```

Comment:
I still don't understand why do we need this `break` 🙂  Can you explain please?

Reply:
given: `expr0, [expr1, expr2, expr3], expr4`. In the first iteration, for example, `input_exprs`(potential extracted exprs) are expr1 and expr3. After expr1 is extracted, it will update loop port, expr2 input have chance to become a loop input port as well, and then could be extracted as well. we should extract expr2 before expr3, so need this break and regenerate input_exprs(potential extracted exprs) as expr2 and expr3 in next iteration. There should be other manner to not break, but this manner I think is simple and safer to extract one by one. What do you think?

Reply:
But we will brake on the next iteration anyway! Because we set `continue_to_extract = false;`

Reply:
L154 means when all exprs are extracted, the loop become a empty loop, then remove this loop and set `continue_to_extract = false;` to finish this loop. In most case, it will not be set on L154 as long as there are still expr in loop.

Reply:
discussed offline, `continue_to_extract = false; `and` break;` have the same functionality, and are redundant, removed one. Thanks Ivan :-)